In [1]:
from mlxtend.frequent_patterns import fpgrowth, association_rules
import pandas as pd
import numpy as np
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

In [7]:
df = pd.read_excel('/home/hady/Studies/Uni/Data Mining/Assignments/Assignment 1/Data/online_retail_II.xlsx')

In [8]:
df.head()

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
0,489434,85048,15CM CHRISTMAS GLASS BALL 20 LIGHTS,12,2009-12-01 07:45:00,6.95,13085.0,United Kingdom
1,489434,79323P,PINK CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
2,489434,79323W,WHITE CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
3,489434,22041,"RECORD FRAME 7"" SINGLE SIZE",48,2009-12-01 07:45:00,2.10,13085.0,United Kingdom
4,489434,21232,STRAWBERRY CERAMIC TRINKET BOX,24,2009-12-01 07:45:00,1.25,13085.0,United Kingdom


In [9]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 525461 entries, 0 to 525460
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   Invoice      525461 non-null  object        
 1   StockCode    525461 non-null  object        
 2   Description  522533 non-null  object        
 3   Quantity     525461 non-null  int64         
 4   InvoiceDate  525461 non-null  datetime64[us]
 5   Price        525461 non-null  float64       
 6   Customer ID  417534 non-null  float64       
 7   Country      525461 non-null  str           
dtypes: datetime64[us](1), float64(2), int64(1), object(3), str(1)
memory usage: 32.1+ MB


In [10]:
df.describe()

,Quantity,InvoiceDate,Price,Customer ID
count,525461.000000,525461,525461.000000,417534.000000
mean,10.337667,2010-06-28 11:37:36.845018,4.688834,15360.645478
min,-9600.000000,2009-12-01 07:45:00,-53594.360000,12346.000000
25%,1.000000,2010-03-21 12:20:00,1.250000,13983.000000
50%,3.000000,2010-07-06 09:51:00,2.100000,15311.000000
75%,10.000000,2010-10-15 12:45:00,4.210000,16799.000000
max,19152.000000,2010-12-09 20:01:00,25111.090000,18287.000000
std,107.424110,NaN,146.126914,1680.811316


# Preprocessing

We will drop rows where Customer ID is null since it won't help us in any way, We will also drop rows in Price and Quantitiy where the value is -ve since it might indicate cancellations or returns or not valid values in general. We will drop InvoiceDate and Country too since they are Irrelevant to what we are doing.

In [11]:
#drop cancellations and returns

df = df[df['Quantity'] > 0]
df = df[df['Price'] > 0]
df = df.dropna(subset=['Customer ID']) #drop rows with missing CustomerID

In [12]:
df.info()

<class 'pandas.DataFrame'>
Index: 407664 entries, 0 to 525460
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   Invoice      407664 non-null  object        
 1   StockCode    407664 non-null  object        
 2   Description  407664 non-null  object        
 3   Quantity     407664 non-null  int64         
 4   InvoiceDate  407664 non-null  datetime64[us]
 5   Price        407664 non-null  float64       
 6   Customer ID  407664 non-null  float64       
 7   Country      407664 non-null  str           
dtypes: datetime64[us](1), float64(2), int64(1), object(3), str(1)
memory usage: 28.0+ MB


In [13]:
df.describe()

,Quantity,InvoiceDate,Price,Customer ID
count,407664.000000,407664,407664.000000,407664.000000
mean,13.585585,2010-07-01 10:15:11.871688,3.294438,15368.592598
min,1.000000,2009-12-01 07:45:00,0.001000,12346.000000
25%,2.000000,2010-03-26 14:01:00,1.250000,13997.000000
50%,5.000000,2010-07-09 15:47:00,1.950000,15321.000000
75%,12.000000,2010-10-14 17:09:00,3.750000,16812.000000
max,19152.000000,2010-12-09 20:01:00,10953.500000,18287.000000
std,96.840747,NaN,34.757965,1679.762138


In [14]:
df = df.drop(['InvoiceDate', 'Country'], axis=1)

In [15]:
df.head()

,Invoice,StockCode,Description,Quantity,Price,Customer ID
0,489434,85048,15CM CHRISTMAS GLASS BALL 20 LIGHTS,12,6.95,13085.0
1,489434,79323P,PINK CHERRY LIGHTS,12,6.75,13085.0
2,489434,79323W,WHITE CHERRY LIGHTS,12,6.75,13085.0
3,489434,22041,"RECORD FRAME 7"" SINGLE SIZE",48,2.10,13085.0
4,489434,21232,STRAWBERRY CERAMIC TRINKET BOX,24,1.25,13085.0


In [20]:
#ensure right data types

df['StockCode'] = df['StockCode'].astype(str)
df['Customer ID'] = df['Customer ID'].astype(int)
df['Description'] = df['Description'].astype(str)
df['Invoice'] = df['Invoice'].astype(int)

In [21]:
df.dtypes

Invoice          int64
StockCode          str
Description        str
Quantity         int64
Price          float64
Customer ID      int64
dtype: object

In [22]:
#dropping duplicate rows

df = df.drop_duplicates()

In [23]:
df.info()

<class 'pandas.DataFrame'>
Index: 400915 entries, 0 to 525460
Data columns (total 6 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   Invoice      400915 non-null  int64  
 1   StockCode    400915 non-null  str    
 2   Description  400915 non-null  str    
 3   Quantity     400915 non-null  int64  
 4   Price        400915 non-null  float64
 5   Customer ID  400915 non-null  int64  
dtypes: float64(1), int64(3), str(2)
memory usage: 21.4 MB


# Getting Rules

## Association Rules

1. Support — how often do these items appear together across ALL transactions
    - Low support = rare combination, High support = very common combination

2. Confidence — given that the customer bought item A, how likely are they to also buy item B
    - Example: conf = 0.7 means 70% of people who bought A also bought B, It's basically a conditional probability

3. Lift — is this association actually meaningful or just a coincidence?
    - Lift = 1 → pure coincidence, the items are independent
    - Lift > 1 → they actually go together more than random chance
    - Lift < 1 → buying A actually makes B less likely
    - We always want lift > 1, otherwise the rule is useless

## Approach

- We will use Invoice and StockCode for Association Rules as they represent a cart or a basket of products
- We will use StockCode and Descripition for TF-IDF to mine for words in descripition per product

In [28]:
basket = df.groupby(['Invoice', 'StockCode'])['Quantity'].sum().unstack().fillna(0)

#groupby → groups by invoice and product
#sum() → adds up quantities
#unstack() → pivots StockCode into columns
#fillna(0) → fills missing combinations with 0

basket = basket.map(lambda x: 1 if x > 0 else 0).astype(bool)

#Converts all numbers into just 0 or 1 (bought or not bought), because association rules don't care about quantity,
#just whether the item was present, converted to bool as fpgrowth algorithm works better with it

In [31]:
# Start with these

min_support = 0.01
min_confidence = 0.5
min_lift = 1.0

In [32]:
#running fp_growth

frequent_itemsets = fpgrowth(basket, min_support, use_colnames=True)

In [34]:
#generating rules

rules = association_rules(frequent_itemsets, metric="confidence", min_threshold=min_confidence)
rules = rules[rules['lift'] >= min_lift]

In [35]:
print(f"Number of rules: {len(rules)}")
print(rules.head())

Number of rules: 244
          antecedents         consequents  antecedent support  \
0  frozenset({22064})  frozenset({21232})            0.016707   
1  frozenset({21755})  frozenset({21754})            0.049394   
2  frozenset({20971})  frozenset({20972})            0.029667   
3  frozenset({22274})  frozenset({22271})            0.018685   
4  frozenset({22274})  frozenset({22273})            0.018685   

   consequent support   support  confidence       lift  representativity  \
0            0.069484  0.010306    0.616822   8.877161               1.0   
1            0.061104  0.027169    0.550053   9.001842               1.0   
2            0.037162  0.018061    0.608772  16.381422               1.0   
3            0.023161  0.010878    0.582173  25.135470               1.0   
4            0.021756  0.010149    0.543175  24.966580               1.0   

   leverage  conviction  zhangs_metric   jaccard  certainty  kulczynski  
0  0.009145    2.428419       0.902429  0.135802   0.5882

# Output Explaination

antecedents: {22064}  →  consequents: {21232}
Means: customers who bought product 22064 also bought product 21232

**The numbers**:

- antecedent support = 0.016707 → Product 22064 appears in 1.67% of all transactions
- consequent support = 0.069484 → Product 21232 appears in 6.94% of all transactions
- support = 0.010306 → Both products appear together in 1.03% of all transactions
- confidence = 0.616822 → 61.7% of customers who bought 22064 also bought 21232
- lift = 8.877 → Customers who bought 22064 are 8.8x more likely to buy 21232 than a random customer. This is a very strong association

**So in Summary**:

If a customer buys product 22064, there's a 61.7% chance they'll also buy product 21232, and this is 8.8 times stronger than random chance